presto-main/src/main/java/io/prestosql/sql/planner/LogicalPlanner.java

```diff
@@ -340,26 +340,26 @@ private RelationPlan createInsertPlan(Analysis analysis, Insert insertStatement)
 
         TableMetadata tableMetadata = metadata.getTableMetadata(session, insert.getTarget());
 
-        List<ColumnMetadata> visibleTableColumns = tableMetadata.getColumns().stream()
-                .filter(column -> !column.isHidden())
-                .collect(toImmutableList());
-        List<String> visibleTableColumnNames = visibleTableColumns.stream()
-                .map(ColumnMetadata::getName)
-                .collect(toImmutableList());
-
         RelationPlan plan = createRelationPlan(analysis, insertStatement.getQuery());
 
         Map<String, ColumnHandle> columns = metadata.getColumnHandles(session, insert.getTarget());
         Assignments.Builder assignments = Assignments.builder();
+        boolean supportsMissingColumnsOnInsert = metadata.supportsMissingColumnsOnInsert(session, insert.getTarget());
+        ImmutableList.Builder<ColumnMetadata> visibleColumnsBuilder = ImmutableList.builder();
+
         for (ColumnMetadata column : tableMetadata.getColumns()) {
             if (column.isHidden()) {
                 continue;
             }
             Symbol output = symbolAllocator.newSymbol(column.getName(), column.getType());
             int index = insert.getColumns().indexOf(columns.get(column.getName()));
             if (index < 0) {
+                if (supportsMissingColumnsOnInsert) {
+                    continue;
+                }
                 Expression cast = new Cast(new NullLiteral(), toSqlType(column.getType()));
                 assignments.put(output, cast);
+                visibleColumnsBuilder.add(column);
```

Comment:
the `visibleColumns(Builder)` name is no longer suitable.
rename to `insertedColumns(Builder)`